integration/tests/params.go

```diff
@@ -0,0 +1,163 @@
+type Params struct {
+	EksctlPath string
+	Region     string
+	Version    string
+	// Flags to help with the development of the integration tests
+	clusterNamePrefix string
+	ClusterName       string
+	DoCreate          bool
+	DoDelete          bool
+	KubeconfigPath    string
+	KubeconfigTemp    bool
+	TestDirectory     string
+	// privateSSHKeyPath is the SSH key to use for Git operations.
+	PrivateSSHKeyPath       string
+	EksctlCmd               runner.Cmd
+	EksctlCreateCmd         runner.Cmd
+	EksctlGetCmd            runner.Cmd
+	EksctlDeleteCmd         runner.Cmd
+	EksctlDeleteClusterCmd  runner.Cmd
+	EksctlScaleNodeGroupCmd runner.Cmd
+	EksctlUtilsCmd          runner.Cmd
+	EksctlExperimentalCmd   runner.Cmd
+	// Keep track of created clusters, for post-tests clean-up.
+	clustersToDelete []string
+}
+
+// SetRegion sets the provided region and re-generates other fields depending
+// on the region field.
+func (p *Params) SetRegion(region string) {
+	p.Region = region
+	p.GenerateCommands()
+}
+
+// GenerateKubeconfigPath generates a path in ${TEMP} based on these params' cluster name.
+func (p *Params) GenerateKubeconfigPath() {
+	p.KubeconfigPath = path.Join(os.TempDir(), fmt.Sprintf("%s.yaml", p.ClusterName))
+}
+
+// GenerateCommands generates eksctl commands with the various options & values
+// provided to this `Params` object.
+func (p *Params) GenerateCommands() {
+	p.EksctlCmd = runner.NewCmd(p.EksctlPath).
+		WithArgs("--region", p.Region).
+		WithTimeout(30 * time.Minute)
+
+	p.EksctlCreateCmd = p.EksctlCmd.
+		WithArgs("create").
+		WithTimeout(25 * time.Minute)
+
+	p.EksctlGetCmd = p.EksctlCmd.
+		WithArgs("get").
+		WithTimeout(1 * time.Minute)
+
+	p.EksctlDeleteCmd = p.EksctlCmd.
+		WithArgs("delete").
+		WithTimeout(15 * time.Minute)
+
+	p.EksctlDeleteClusterCmd = p.EksctlDeleteCmd.
+		WithArgs("cluster", "--verbose", "4")
+
+	p.EksctlScaleNodeGroupCmd = p.EksctlCmd.
+		WithArgs("scale", "nodegroup", "--verbose", "4").
+		WithTimeout(5 * time.Minute)
+
+	p.EksctlUtilsCmd = p.EksctlCmd.
+		WithArgs("utils").
+		WithTimeout(5 * time.Minute)
+
+	p.EksctlExperimentalCmd = p.EksctlCmd.
+		WithEnv("EKSCTL_EXPERIMENTAL=true")
+}
+
+// NewClusterName generates a new cluster name using the provided prefix, and
+// adds the cluster to the list of clusters to eventually delete, once the test
+// suite has run.
+func (p *Params) NewClusterName(prefix string) string {
+	return p.formatClusterName(prefix, names.ForCluster("", ""))
+}
+
+// FormatClusterName creates a new cluster name using this Params struct's
+// cluster name prefix, the provided name, and adds the cluster to the list of
+// clusters to eventually delete, once the test suite has run.
+func (p *Params) FormatClusterName(name string) string {
+	return p.formatClusterName(p.clusterNamePrefix, name)
+}
+
+func (p *Params) formatClusterName(prefix string, name string) string {
+	clusterName := fmt.Sprintf("it-%s-%s", prefix, name)
+	p.addToDeleteList(clusterName)
```

Comment:
Can we move the addition to the list to line 96?

Reply:
We could, but we need to expose some extra method to add a cluster to the list of clusters to delete, and call that method from somewhere in the test. (I actually don't really like the way this all works, but haven't really invested time into making it better. For now, I wanted it to roughly work, and to focus on parallelising things.)

Reply:
I'm not sure I understand why. What I mean is something like this:

```func (p *Params) NewClusterName(prefix string) string {
	formattedName := p.formatClusterName(prefix, names.ForCluster("", ""))
        p.addToDeleteList(clusterName)
        return formattedName
}

func (p *Params) FormatClusterName(name string) string {
	return p.formatClusterName(p.clusterNamePrefix, name)
}

func (p *Params) formatClusterName(prefix string, name string) string {
	clusterName := fmt.Sprintf("it-%s-%s", prefix, name)
	return clusterName
}
```
So that `FormatClusterName()` does only one thing. Right now it says format name but it has a side effect of changing the state of `p` without the caller knowing unless they read it's comment.

Then a caller can a) only format a name or b) format the name and add it to the list.

Reply:
> I'm not sure I understand why.

I agree and also usually tend to have methods which do "one thing" and have no side effect, but in this specific case, I wanted to avoid "polluting" the integration tests, to avoid having extra "test infrastructure" related calls in them, as it makes it harder to understand the essence of what they do, of what they test, so was happy with that ~side-effect~ "magic".